src/Microsoft.Health.Fhir.Liquid.Converter/DotLiquids/MergeDiff.cs

```diff
@@ -0,0 +1,135 @@
+        }
+
+        public override void Render(Context context, TextWriter result)
+        {
+
+            context.Stack(() =>
+            {
+                using StringWriter writer = new StringWriter();
+
+                var inputContent = context[_variableName];
+
+                // If input message is null, it returns empty string to makesure the output is still a valid json object.
+                // And the element with empty string value will be removed in post processor.
+                if (inputContent == null)
+                {
+                    result.Write("\"\"");
+                    return;
+                }
+
+                // Input message should be a valid json object.
+                if (!(inputContent is Dictionary<string, object> || inputContent is Hash))
+                {
+                    throw new RenderException(FhirConverterErrorCode.InvalidInputOfMergeDiffBlock, "The input message of MergeDiff tag should be a json object.");
+                }
+
+                RenderAll(DiffBlock, context, writer);
+
+                // Diff content should be a valid json object.
+                // If diff content is empty, the input message will output directly.
+                Dictionary<string, object> diffDic = null;
+                try
+                {
+                    diffDic = JsonConvert.DeserializeObject<Dictionary<string, object>>(writer.ToString());
+                }
+                catch
+                {
+                    throw new RenderException(FhirConverterErrorCode.InvalidMergeDiffBlockContent, "MergeDiff block content should be a json object.");
+                }
+
+                var mergedResult = inputContent;
+                if (diffDic != null)
+                {
+                    mergedResult = MergeDiffContent(inputContent, diffDic);
+                }
+
+                result.Write(mergedResult == null ? null : JsonConvert.SerializeObject(mergedResult));
+            });
+        }
+
+        public object MergeDiffContent(object source, Dictionary<string, object> diffDic)
+        {
+            Dictionary<string, object> result;
+
+            if (source is Hash)
+            {
+                result = new Dictionary<string, object>(source as Hash);
+            }
+            else if (source is Dictionary<string, object>)
+            {
+                result = new Dictionary<string, object>(source as Dictionary<string, object>);
+            }
+            else
+            {
+                throw new RenderException(FhirConverterErrorCode.InvalidInputOfMergeDiffBlock, "MergeDiff block is empty.");
+            }
+
+            foreach (var item in diffDic)
+            {
+                if (item.Key.EndsWith("[x]"))
+                {
+                    var choiceTypeName = item.Key.Replace("[x]", string.Empty);
+                    var choiceElements = result.Where(x => x.Key.StartsWith(choiceTypeName)).Select(x => x.Key).ToList();
```

Comment:
I'm afraid if x.Key.StartsWith() is sufficient for finding a choice type, for example, if there are `onsetDatetime` and `onsetXXX` properties, shall we also check the type field to ensure the property key is choice type as the number of qualified types are quite limited?

Reply:
After offline discussion, we will give detailed explanation for the usage of [x] in document.